app/coffee/DocumentUpdaterController.coffee

```diff
@@ -74,9 +74,7 @@ module.exports = DocumentUpdaterController =
 			if err?
 				return logger.err {room: doc_id, err}, "failed to get room clients"
 
-			for clientId in clientIds
-				continue unless io.sockets.connected.hasOwnProperty(clientId)
-				client = io.sockets.connected[clientId]
+			for client in clientIds.map((id) -> io.sockets.connected[id]) when client # filter disconnected clients
```

Comment:
What actually happens if we emit to or disconnect a client that has disconnected?  

Maybe this is just how it works but I feel like there should be some way to handle this without having to check clients explicitly.

Reply:
Here we are just receiving the list of client ids that were connected at the time of the previous process tick.

https://github.com/socketio/socket.io/blob/47161a65d40c2587535de750ac4c7d448e5842ba/lib/socket.js#L453
The disconnect logic removes the client from the `io.sockets.connected` Object immediately and we would try to call `undefined.emit(..)`.

Reply:
Thanks for the link.   As I understand it `process.nextTick` will run after the current javascript finishes executing but _before_ the next event loop, so it is effectively a synchronous callback.  Does that make a difference? I think that would make the clients guaranteed to be valid and we could skip the check.

https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/#process-nexttick

Reply:
TIL, thanks. I will remove the checks then.

Reply:
[I've reworked the frontend code](https://github.com/overleaf/web-internal/compare/59d3ba0b4344734e323cfbd1bcb29618d5bd7b68..169d90cceb8f7e618c2ceb17da612d39519ef4c7) -- and was bitten by the early client cleanup :disappointed: 

Anyways here is the patch 8026003f8e9dab506d1a8a3e6e5f41188555a6bb.